nixos/modules/flake/source-info.nix

```diff
@@ -0,0 +1,31 @@
+{ config, lib, ... }:
+
+with lib;
+
+let sourceInfo = config.system.nixos.configuration.sourceInfo;
+
+in
+{
+  options = {
+
+    system.nixos.configuration.sourceInfo = mkOption {
+      type = types.attrsOf types.anything;
+      default = {};
+      example = "self.sourceInfo";
+      description = ''
+        The source information of a flake-based NixOS configuration.
+        If set, the attribute <literal>configurationRevision</literal>
+        will appear properly in the output of
+        <literal>nixos-version --json</literal>.
```

Comment:
Doesn't it also occur in the grub and systemd-boot boot menus?

Reply:
It does not. Currently, it contains [only the `shortRev` of `nixpkgs`](https://github.com/NixOS/nixpkgs/blob/0b683abff06fe55755ea992ba47f2e787081a30f/flake.nix#L26). Should we add the configuration revision to `versionSuffix` too? If both revisions are to be included, they should be made distinctive.

Note that the `nixpkgs` revision can be `dirty` too; we may develop a uniform way to obtain revisions, and apply it to both occasions. What do you think about my earlier proposals on setting the `dirty` tag? I can start to work on a prototype after we discuss and reach an agreement.

Reply:
> It does not.

Ok. That can be done in another PR if desirable.



> we may develop a uniform way to obtain revisions

Sounds good, but these are all quite independent problems. It's best to resolve them in separate PRs, so that maintainers don't need to keep (in the limit) all of NixOS in their head in order to approve them.

Reply:
Sure, there will be at least three PRs: this one, one for obtaining revisions, and another one for applying the feature. I will try to write a detailed design later.